File: apps/Offline-AI-Buddy/shared/BuddyAICore/Sources/BuddyAICore/LLM/ModelStore.swift
import Foundation

/// On-disk layout for the GGUF model file. Single file in v1; v1.1's
/// model marketplace will use the same `models/` directory with one
/// file per manifest entry.
public final class ModelStore {

    public let directory: URL
    private let fileManager: FileManager

    public init(directory: URL? = nil, fileManager: FileManager = .default) throws {
        self.fileManager = fileManager
        let base: URL
        if let directory {
            base = directory
        } else {
            base = try fileManager
                .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
                .appendingPathComponent("offlineaibuddy", isDirectory: true)
                .appendingPathComponent("models", isDirectory: true)
        }
        try fileManager.createDirectory(at: base, withIntermediateDirectories: true)
        self.directory = base
    }

    public func url(forModelNamed name: String) -> URL {
        directory.appendingPathComponent(name)
    }

    public func isInstalled(named name: String) -> Bool {
        fileManager.fileExists(atPath: url(forModelNamed: name).path)
    }

    /// Move a downloaded file into the store. Atomic replace.
    @discardableResult
    public func install(downloaded source: URL, named name: String) throws -> URL {
        let dest = url(forModelNamed: name)
        if fileManager.fileExists(atPath: dest.path) {
            try fileManager.removeItem(at: dest)
        }
        try fileManager.moveItem(at: source, to: dest)
        return dest
    }

    public func remove(named name: String) throws {
        let target = url(forModelNamed: name)
        if fileManager.fileExists(atPath: target.path) {
            try fileManager.removeItem(at: target)
        }
    }

    /// Verify the SHA-256 of an installed model. `expected` of `""`
    /// counts as "skip verification" (dev mode).
    public func verify(named name: String, expectedSHA256 expected: String) -> Bool {
        if expected.isEmpty { return true }
        let actual = ModelDownloader.sha256(of: url(forModelNamed: name))
        return actual.lowercased() == expected.lowercased()
    }
}

File: apps/Offline-AI-Buddy/android/core/src/main/java/com/americangroupllc/offlineaibuddy/core/storage/ModelStore.kt
package com.americangroupllc.offlineaibuddy.core.storage

import java.io.File
import java.security.MessageDigest

/**
 * On-disk layout for the GGUF model file. Mirrors `BuddyAICore.ModelStore`.
 */
class ModelStore(directory: File) {

    val directory: File = File(directory, "models").also { it.mkdirs() }

    fun urlFor(name: String): File = File(directory, name)

    fun isInstalled(name: String): Boolean = urlFor(name).exists()

    /** Move a downloaded file into the store. Atomic replace. */
    fun install(downloaded: File, name: String): File {
        val dest = urlFor(name)
        if (dest.exists()) dest.delete()
        if (!downloaded.renameTo(dest)) {
            downloaded.copyTo(dest, overwrite = true)
            downloaded.delete()
        }
        return dest
    }

    fun remove(name: String) {
        val target = urlFor(name)
        if (target.exists()) target.delete()
    }

    /** SHA-256 hex digest of an installed model. `expectedSha256 = ""` skips. */
    fun verify(name: String, expectedSha256: String): Boolean {
        if (expectedSha256.isEmpty()) return true
        val target = urlFor(name)
        if (!target.exists()) return false
        val digest = MessageDigest.getInstance("SHA-256")
        target.inputStream().use { input ->
            val buf = ByteArray(1024 * 1024)
            while (true) {
                val read = input.read(buf)
                if (read <= 0) break
                digest.update(buf, 0, read)
            }
        }
        val actual = digest.digest().joinToString("") { "%02x".format(it) }
        return actual.equals(expectedSha256, ignoreCase = true)
    }
}
